data_structures/rtree.hpp

```diff
@@ -64,14 +65,135 @@ namespace spatial {
                     std::shared_ptr<Datum<T>> get_datum() const { 
                         return std::get<std::shared_ptr<Datum<T>>>(
                             _contents
-                        ); 
-                    }
+                        );
+                }
 
                     std::shared_ptr<Node> get_node() const { 
                         return std::get<std::shared_ptr<Node>>(
                             _contents
-                        ); 
+                        );
+                    }
+
+                    bool is_leaf_entry() const {
+                        return std::holds_alternative
+                            <std::shared_ptr<Datum<T>>>(
+                                _contents
+                            );
+                    }
+            };
+
+            /**
+             * A thin wrapper around a std::priority_queue for Entries.
+             * Helps to simplify the distance browsing code.
+             */
+            class EntryPQ {
+                private:
+                    struct EntryPQE {
+                        Entry entry;
+                        coord_t dist;
+                    };
+
+                    struct Closer {
+                        bool operator()(EntryPQE const a, EntryPQE const b) {
+                            return (a.dist > b.dist);
+                        }
+                    };
```

Comment:
```suggestion
            template < typename S >
            class knnPQ {
                private:
                    struct PQE {
                        S entry;
                        coord_t dist;
                    };

                    struct Farther {
                        bool operator()(PQE const a, PQE const b) const {
                            return (a.dist > b.dist);
                        }
                    };
```
I don't expect you to make this change (and its cascading effects), just illustrating how you could templatise this to reduce code duplication (and conceivably reuse it elsewhere), using `S` so that it doesn't "shadow" the existing template parameter `T`. Usage in the query algorithm would then be:
```
    knnPQ<Entry> entry_pq(query_point);
    entry_pq.push(*root_entry);
    knnPQ<Datum<T>> datum_pq(query_point);
``` 

_Note the constness added to the functor_. Considering how often this functor gets called, you could lose quite a bit of performance in theory (though I suspect it would compile the same anyway).